### PR TITLE
[ExtraShotLatte] iP

### PR DESCRIPTION
**Pull request by ExtraShotLatte**
DukeNotSoPro makes you try to remember things you have not done.

_It's:_
1. text-based
2. easy to **learn**
3. **FAST** _SUPER FAST_ to use
4. All ~you~ need to do is,

download it from [here](https://nus-cs2103-ay2223s2.github.io/website/schedule/week4/project.html).
- double-click it.
- add your tasks.
- let it manage your tasks for you 😉
- And it is FREE!

Features:
- [ ] Managing tasks
- [x]  Managing deadlines (coming soon)
- [x]  Reminders (coming soon)

If you Java programmer, you can use it to practice Java too. Here's the `main` method:

```java
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```

### CODE REVIEW

File: src/main/java/duke/Deadline.java [hunk 1]
```diff
@@ -0,0 +1,65 @@
+package duke;
+
+import java.time.LocalDate;
+import java.time.format.DateTimeFormatter;
+
+/**
+ * Models an deadline which is a task.
+ */
+public class Deadline extends Task {
```
Comment (leonpoh) [inline src/main/java/duke/Deadline.java:9]: Looks good seperating task into its subclasses

**Review by leonpoh — COMMENTED** [review]
I like that you abstracted out the Task, however, maybe you should create another class that executes those command.

File: src/main/java/Duke.java [hunk 2]
```diff
@@ -1,10 +1,66 @@
...
-        String logo = " ____        _        \n"
-                + "|  _ \\ _   _| | _____ \n"
-                + "| | | | | | | |/ / _ \\\n"
-                + "| |_| | |_| |   <  __/\n"
-                + "|____/ \\__,_|_|\\_\\___|\n";
-        System.out.println("Hello from\n" + logo);
+
+        new Duke("tasks.txt").runDuke();
     }
 }
```
Comment (shenchenzizoe) [inline src/main/java/Duke.java:54]: nice OOP implemented

File: src/main/java/duke/Deadline.java [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+     *
+     * @param desc The name of the task.
+     * @param by The deadline of the task.
+     * @param b The status of the task.
+     */
+    public Deadline(String desc, String by, boolean b) {
+        super(desc, b);
+        this.desc = desc;
+        this.by = by;
+        byDate = LocalDate.parse(by);
```
Comment (shenchenzizoe) [inline src/main/java/duke/Deadline.java]: could consider handle exceptions when String by does not meet the LocalDate format?

File: src/main/java/duke/Parser.java [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+            tasks.printNewestTask();
+            storage.save(tasks);
+            break;
+        case "mark":
+            selectedNum = Integer.parseInt(inputCmdArr[1]);
+            tasks.markTask(selectedNum);
+            storage.save(tasks);
+            break;
+        case "unmark":
+            selectedNum = Integer.parseInt(inputCmdArr[1]);
```
Comment (ryanchua00) [inline src/main/java/duke/Parser.java]: Would this throw a NumberFormatException?

File: src/main/java/duke/Deadline.java [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public String toString() {
+        return "[D]" + super.toString() + " (by: " + byDate.format(DateTimeFormatter.ofPattern("MMM d yyyy")) + ")";
```
Comment (shenchenzizoe) [inline src/main/java/duke/Deadline.java:52]: splitting into two lines might meet the coding standards better

File: src/main/java/duke/Storage.java [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+            } catch (IOException e) {
+                throw new RuntimeException(e);
+            }
+        } else {
+            System.out.println("NO EXIST");
+            try {
+                FileUtils.write(dukeDataFile, "");
+            } catch (IOException e) {
+                throw new RuntimeException(e);
+            }
```
Comment (ryanchua00) [inline src/main/java/duke/Storage.java]: Can this try-catch block be moved outside of the if-else?

File: src/main/java/duke/Task.java [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+            return "[ ] " + name;
+        }
+    }
+
+    /**
+     * Method to return the CSV representation of the task.
+     *
+     * @return Task in CSV representation.
+     */
+    public abstract String asCsv();
```
Comment (ryanchua00) [inline src/main/java/duke/Task.java:66]: I like this! 
Although it does necessitate storing "desc" as a field in every child class.

File: src/main/java/duke/Task.java [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+     * Method to mark task as done.
+     */
+    public void markDone() {
+        isDone = true;
+    }
+
+    /**
+     * Method to unmark task as done.
+     */
+    public void markUndone() {
```
Comment (shenchenzizoe) [inline src/main/java/duke/Task.java:44]: may fit the required command better if change method name to "mark & unmark"

**Review by ryanchua00 — COMMENTED** [review]
Overall, great code. It was easy to read and understand. I also understand that with the iP we code more to make it work and have little to no time to go back to improve things. So with the things I picked out, they are just small optimizations for the most part.

**Review by shenchenzizoe — COMMENTED** [review]
Generally the appllication of OOP, the use of packages, and the use of Enums makes the code structured and quite readable! Just some minor coding standard issues to notice.

**Pull request closed without merging**